File: crates/sdk/bindings/assemblyscript/assembly/procedure.ts
// mp2 byte-pipe helpers for the mtp-generated procedure adapters: decode the
// `UniProcedureParam` the runtime passes to each `mp2-<proc>` export and
// encode the `UniResult<UniProcedureResult, UniError>` it expects back (a
// single-entry MessagePack map: key 0 = ok, key 1 = `UniError`).

import { MpackReader, MpackWriter } from "./mpack";
import { UniDataValue } from "./generated/UniDataValue";
import { UniError, UniErrorCodec } from "./generated/UniError";
import { UniProcedureParam, UniProcedureParamCodec } from "./generated/UniProcedureParam";
import { UniProcedureResult, UniProcedureResultCodec } from "./generated/UniProcedureResult";
import { ERROR_INTERNAL, MuduError, alloc, liftBytes } from "./wit";
import { ValueList } from "./sql";

// Decode the canonical-ABI argument: `paramLen` MessagePack bytes at
// `paramPtr`, placed in guest memory by the runtime through `cabi_realloc`.
export function decodeProcedureParam(paramPtr: usize, paramLen: usize): UniProcedureParam {
  const bytes = liftBytes(paramPtr, paramLen);
  const reader = new MpackReader(bytes);
  return UniProcedureParamCodec.decode(reader);
}

// Ok arm: `{0: {1: [return_list]}}`.
export function encodeProcedureOk(values: ValueList): usize {
  return encodeProcedureOkUni(values.toUniDataValues());
}

// Ok arm over raw wire datums: the shape record-typed results take (a
// `ValueList` carries only the scalar DX value kinds, so a record result is
// wrapped by `recordFromFieldValues` and passed here directly).
export function encodeProcedureOkUni(values: Array<UniDataValue>): usize {
  const result = new UniProcedureResult();
  result.return_list = values;
  const writer = new MpackWriter();
  writer.writeMapHeader(1);
  writer.writeU64(0);
  UniProcedureResultCodec.encode(result, writer);
  return returnBytes(writer.toBytes());
}

// Err arm: `{1: UniError}`; the error source defaults to "assemblyscript" and
// the error location to the procedure name (mirroring `procedureResultErr`).
export function encodeProcedureErr(error: MuduError, procedure: string): usize {
  const uniError = new UniError();
  uniError.err_code = error.code == 0 ? ERROR_INTERNAL : error.code;
  uniError.err_msg = error.message;
  uniError.err_src = error.source.length > 0 ? error.source : "assemblyscript";
  uniError.err_loc = error.location.length > 0 ? error.location : procedure;
  const writer = new MpackWriter();
  writer.writeMapHeader(1);
  writer.writeU64(1);
  UniErrorCodec.encode(uniError, writer);
  return returnBytes(writer.toBytes());
}

// A `func(...) -> list<u8>` export returns its payload through a pointer to a
// (ptr, len) pair in guest memory.
function returnBytes(bytes: Uint8Array): usize {
  const pair = alloc(8);
  store<u32>(pair, <u32>(changetype<usize>(bytes.buffer) + bytes.byteOffset));
  store<u32>(pair + 4, <u32>bytes.length);
  return pair;
}

File: crates/sdk/bindings/python/mududb/codec/procedure.py
"""mp2 byte-pipe helpers for procedure adapters (`mududb.codec.procedure`).

Decodes the `UniProcedureParam` the runtime passes to each `mp2-<proc>`
export and encodes the `UniResult<UniProcedureResult, UniError>` it expects
back (a single-entry MessagePack map: key 0 = ok, key 1 = `UniError`).
Mirrors the AssemblyScript `procedure.ts` helpers exactly.
"""

from mududb.codec.mpack import MpackReader, MpackWriter
from mududb.generated.uni_data_value import UniDataValue
from mududb.generated.uni_error import UniError, uni_error_to_value
from mududb.generated.uni_procedure_param import (
    UniProcedureParam,
    uni_procedure_param_from_value,
)
from mududb.generated.uni_procedure_result import (
    UniProcedureResult,
    uni_procedure_result_to_value,
)

__all__ = [
    "decode_procedure_param",
    "encode_procedure_ok",
    "encode_procedure_err",
]

ERROR_INTERNAL = 1


def decode_procedure_param(param: bytes) -> UniProcedureParam:
    """Decode the canonical-ABI argument bytes of an `mp2-<proc>` export."""
    reader = MpackReader(param)
    return uni_procedure_param_from_value(reader.read_value())


def _encode_arm(key: int, value) -> bytes:
    writer = MpackWriter()
    writer.write_map_header(1)
    writer.write_u64(key)
    writer.write_value(value)
    return writer.to_bytes()


def encode_procedure_ok(values: list[UniDataValue]) -> bytes:
    """Ok arm: `{0: {1: [return_list]}}`."""
    return _encode_arm(
        0,
        uni_procedure_result_to_value(UniProcedureResult(return_list=values)),
    )


def encode_procedure_err(
    code: int, message: str, source: str = "python", location: str = ""
) -> bytes:
    """Err arm: `{1: UniError}`; the error source defaults to "python"."""
    error = UniError(
        err_code=code if code != 0 else ERROR_INTERNAL,
        err_msg=message,
        err_src=source,
        err_loc=location,
    )
    return _encode_arm(1, uni_error_to_value(error))
